### PR TITLE
Move manage integrations button from settings page to room header as a stand-alone component

### PR DESCRIPTION
**Pull request by rxl881**
The intent of this PR is to move the existing manage integrations button out from the settings page in to a reusable component, so that it can be used in the room header (and other places, as needed).

### CODE REVIEW

File: src/components/views/elements/ManageIntegsButton.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2017 Vector Creations Ltd
```
Comment (lukebarnard1) [inline src/components/views/elements/ManageIntegsButton.js]: Not anymore!

File: src/components/views/elements/ManageIntegsButton.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+import { _t } from '../../../languageHandler';
+import AccessibleButton from './AccessibleButton';
+import TintableSvg from './TintableSvg';
+
+export default class ManageIntegsButton extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            scalar_error: null,
```
Comment (lukebarnard1) [inline src/components/views/elements/ManageIntegsButton.js]: Should be camelCase

File: src/components/views/elements/ManageIntegsButton.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            showIntegrationsError: !this.state.showIntegrationsError,
+        });
+    }
+
+    render() {
+        let integrationsButton;
+        let integrationsError;
+        if (this.scalarClient !== null) {
+            if (this.state.showIntegrationsError && this.state.scalar_error) {
+                console.error(this.state.scalar_error);
```
Comment (lukebarnard1) [inline src/components/views/elements/ManageIntegsButton.js]: This should probably be logged after line 50

File: src/components/views/elements/ManageIntegsButton.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+        this.onManageIntegrations = this.onManageIntegrations.bind(this);
+        this.onShowIntegrationsError = this.onShowIntegrationsError.bind(this);
+    }
+
+    componentWillMount() {
+        ScalarMessaging.startListening();
+        this.scalarClient = null;
+
+        if (SdkConfig.get().integrations_ui_url && SdkConfig.get().integrations_rest_url) {
```
Comment (lukebarnard1) [inline src/components/views/elements/ManageIntegsButton.js]: It's strange that we don't pass these URLs into the auth client here, and have to check the config before creating a ScakarAuthClient.

File: src/components/views/elements/ManageIntegsButton.js [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        ScalarMessaging.stopListening();
+    }
+
+    onManageIntegrations(ev) {
+        ev.preventDefault();
+        const IntegrationsManager = sdk.getComponent("views.settings.IntegrationsManager");
+        Modal.createDialog(IntegrationsManager, {
+            src: (this.scalarClient !== null && this.scalarClient.hasCredentials()) ?
+                this.scalarClient.getScalarInterfaceUrlForRoom(this.props.room.roomId) :
+                null,
```
Comment (lukebarnard1) [inline src/components/views/elements/ManageIntegsButton.js]: Would this not open an iFrame with a `null` src?

File: src/components/views/elements/ManageIntegsButton.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+                );
+            }
+        }
+
+        return integrationsButton;
+    }
+}
+
+ManageIntegsButton.propTypes = {
+    room: PropTypes.object.isRequired,
```
Comment (lukebarnard1) [inline src/components/views/elements/ManageIntegsButton.js]: I'd prefer a `roomId` going in here, `ManageIntegsButton` doesn't need the rest of room state.

File: src/components/views/elements/ManageIntegsButton.js [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    onManageIntegrations(ev) {
+        ev.preventDefault();
+        const IntegrationsManager = sdk.getComponent("views.settings.IntegrationsManager");
+        Modal.createDialog(IntegrationsManager, {
+            src: (this.scalarClient !== null && this.scalarClient.hasCredentials()) ?
+                this.scalarClient.getScalarInterfaceUrlForRoom(this.props.room.roomId) :
+                null,
+            onFinished: ()=>{
+                if (this.props.onCancelClick) {
+                    this.props.onCancelClick(ev);
```
Comment (lukebarnard1) [inline src/components/views/elements/ManageIntegsButton.js]: You probably don't want to be returning this `ev` here

File: src/components/views/rooms/RoomHeader.js [hunk 8]
```diff
@@ -320,10 +323,15 @@ module.exports = React.createClass({
         }
 
         let rightRow;
+
         if (!this.props.editing) {
             rightRow =
                 <div className="mx_RoomHeader_rightRow">
                     { settingsButton }
+                    <ManageIntegsButton
+                        onCancelClick={this.props.onCancelClick}
```
Comment (lukebarnard1) [inline src/components/views/rooms/RoomHeader.js]: It seems _quite_ bizarre to be bubbling the `onCancelClick` of the integrations manager to the `onCancelClick` of the room header. Stranger still that the `onCancelClick` that is passed to RoomView will close room settings, so this is definitely wrong.

File: src/components/views/elements/ManageIntegsButton.js (hunk 4 above)
Comment (rxl881) [inline src/components/views/elements/ManageIntegsButton.js]: Not an argument against fixing it, but this is simply copied across from the existing code on the settings page -- https://github.com/matrix-org/matrix-react-sdk/pull/1286/files#diff-7810353d2ee5218b7242e81423030433L141.

Unless you think this is broken, I'm inclined to leave it. What do you reckon?

File: src/components/views/elements/ManageIntegsButton.js (hunk 7 above)
Comment (rxl881) [inline src/components/views/elements/ManageIntegsButton.js]: As with one of the previous comments, this code is directly copied across from the existing settings page - https://github.com/matrix-org/matrix-react-sdk/pull/1286/files#diff-7810353d2ee5218b7242e81423030433L525.

I'm inclined to leave it, for fear of breaking existing expected behaviour.

File: src/components/views/elements/ManageIntegsButton.js (hunk 6 above)
Comment (rxl881) [inline src/components/views/elements/ManageIntegsButton.js]: Done.

File: src/components/views/rooms/RoomHeader.js (hunk 8 above)
Comment (rxl881) [inline src/components/views/rooms/RoomHeader.js]: Thanks for spotting that!

**Comment by rxl881** [comment]
Thanks Luke. Hopefully that is mainly sorted now. There are two items that were copied across from existing code and I'm tempted to leave, unless you think that they should be addressed now. Can you PTAL?

Cheers.

File: src/components/views/elements/ManageIntegsButton.js (hunk 5 above)
Comment (rxl881) [inline src/components/views/elements/ManageIntegsButton.js]: As with some of the previous comments, this has just been cargo-culted across from the existing room settings code (https://github.com/matrix-org/matrix-react-sdk/pull/1286/files#diff-7810353d2ee5218b7242e81423030433L520). As it is not currently considered broken, I'm inclined to leave it as is, for now.

File: src/components/views/elements/ManageIntegsButton.js (hunk 4 above)
Comment (lukebarnard1) [inline src/components/views/elements/ManageIntegsButton.js]: It's not broken, it's just a weird API that made sense when we only ever made one Scalar client. Feel free to leave it for now.

File: src/components/views/elements/ManageIntegsButton.js [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    onManageIntegrations(ev) {
+        ev.preventDefault();
+        const IntegrationsManager = sdk.getComponent("views.settings.IntegrationsManager");
+        Modal.createDialog(IntegrationsManager, {
+            src: (this.scalarClient !== null && this.scalarClient.hasCredentials()) ?
+                this.scalarClient.getScalarInterfaceUrlForRoom(this.props.roomId) :
+                null,
+            onFinished: ()=>{
+                if (this.props.onCancelClick) {
+                    this.props.onCancelClick(ev);
```
Comment (lukebarnard1) [inline src/components/views/elements/ManageIntegsButton.js]: > for fear of breaking existing expected behaviour.

This property is never set - the entire `if` is redundant and we should absolutely avoid redundant, potentially confusing code.

This means that the behaviour has changed, given you've copied it. Oh! I guess this is the thing where if you close the integrations manager, you also close the room settings. Maybe that wasn't so bizarre after all?

File: src/components/views/elements/ManageIntegsButton.js [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+            showIntegrationsError: !this.state.showIntegrationsError,
+        });
+    }
+
+    render() {
+        let integrationsButton;
+        let integrationsError;
+        if (this.scalarClient !== null) {
+            if (this.state.showIntegrationsError && this.state.scalarError) {
+                console.error(this.state.scalarError);
```
Comment (lukebarnard1) [inline src/components/views/elements/ManageIntegsButton.js]: I suggest moving this to where the error is first caught.

File: src/components/views/rooms/RoomHeader.js [hunk 11]
```diff
@@ -47,13 +48,15 @@ module.exports = React.createClass({
         onSaveClick: React.PropTypes.func,
         onSearchClick: React.PropTypes.func,
         onLeaveClick: React.PropTypes.func,
+        onCancelClick: React.PropTypes.func,
```
Comment (lukebarnard1) [inline src/components/views/rooms/RoomHeader.js]: `onCancelClick` is not used by anything

File: src/components/views/elements/ManageIntegsButton.js (hunk 10 above)
Comment (rxl881) [inline src/components/views/elements/ManageIntegsButton.js]: Done.

File: src/components/views/elements/ManageIntegsButton.js (hunk 9 above)
Comment (rxl881) [inline src/components/views/elements/ManageIntegsButton.js]: As discussed IRL, I've removed this all together from the ManageIntegsButton component, as it doesn't appear to be used at the moment.

File: src/components/views/rooms/RoomHeader.js (hunk 11 above)
Comment (rxl881) [inline src/components/views/rooms/RoomHeader.js]: Although it is not used by anything that I have added, this prop actually is used elsewhere in the code... https://github.com/matrix-org/matrix-react-sdk/pull/1286/files#diff-96b79748dd7ebec5d36339d2478e92cbR194

I was adding it in to the component props and default to make it clear to future people (like me) that this prop is passed in are available to be used by other components and to make sure that there is a sensible default if the function is not passed.

I can remove it if you want, but I think it makes more sense to leave it?

File: src/components/views/elements/ManageIntegsButton.js (hunk 4 above)
Comment (rxl881) [inline src/components/views/elements/ManageIntegsButton.js]: Cool. Cheers.

File: src/components/views/elements/ManageIntegsButton.js [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+        return integrationsButton;
+    }
+}
+
+ManageIntegsButton.propTypes = {
+    roomId: PropTypes.string.isRequired,
+};
+
+ManageIntegsButton.defaultProps = {
```
Comment (lukebarnard1) [inline src/components/views/elements/ManageIntegsButton.js]: Empty default props

File: src/components/views/rooms/RoomHeader.js (hunk 11 above)
Comment (lukebarnard1) [inline src/components/views/rooms/RoomHeader.js]: My mistake, this looks to be necessary. I think I was confused by the fact that it wasn't in prop types previously.

File: src/components/views/elements/ManageIntegsButton.js (hunk 4 above)
Comment (t3chguy) [inline src/components/views/elements/ManageIntegsButton.js]: I wrote this check to hide all instances of Manage Integrations/Scalar/Modular if there is are no integrations URLs defined in the config. i.e a way to hide the buttons.

This seems to have regressed from the original function: https://github.com/matrix-org/matrix-react-sdk/pull/816

**Pull request merged**